x/coredaos/keeper/msg_server.go

```diff
@@ -0,0 +1,409 @@
+		"new_end_time", proposal.VotingEndTime,
+		"authority", msg.Extender,
+		"times_extended", proposal.TimesVotingPeriodExtended,
+	)
+
+	// Emit event for voting period extension
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeExtendVotingPeriod,
+			sdk.NewAttribute(types.AttributeKeyProposalID, fmt.Sprintf("%d", proposal.Id)),
+			sdk.NewAttribute(types.AttributeKeySigner, msg.Extender),
+			sdk.NewAttribute(types.AttributeKeyNewEndTime, proposal.VotingEndTime.String()),
+			sdk.NewAttribute(types.AttributeKeyTimesExtended, fmt.Sprintf("%d", proposal.TimesVotingPeriodExtended)),
+		),
+	})
+
+	return &types.MsgExtendVotingPeriodResponse{}, nil
+}
+
+// VetoProposal allows the signer to veto a proposal.
+// The proposal must be in the voting period, and the signer must be the designated Oversight DAO.
+// If the proposal is vetoed, it will be removed from the active proposals queue and rejected.
+func (ms MsgServer) VetoProposal(goCtx context.Context, msg *types.MsgVetoProposal) (*types.MsgVetoProposalResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	params := ms.k.GetParams(ctx)
+
+	logger := ms.k.Logger(ctx)
+
+	if params.OversightDaoAddress == "" {
+		logger.Info("Oversight DAO address is not set, function is disabled")
+
+		return nil, errors.Wrapf(types.ErrFunctionDisabled, "Oversight DAO address is not set")
+	}
+
+	if msg.Vetoer != params.OversightDaoAddress {
+		logger.Error(
+			"invalid authority for vetoing proposal",
+			"expected", params.OversightDaoAddress,
+			"got", msg.Vetoer,
+		)
+
+		return nil, errors.Wrapf(types.ErrInvalidSigner, "invalid authority; expected %s, got %s", params.OversightDaoAddress, msg.Vetoer)
+	}
+
+	proposal, found := ms.k.govKeeper.GetProposal(ctx, msg.ProposalId)
+	if !found {
+		logger.Error(
+			"proposal not found",
+			"proposal_id", msg.ProposalId,
+			"authority", msg.Vetoer,
+		)
+
+		return nil, errors.Wrapf(govtypes.ErrUnknownProposal, "proposal with ID %d not found", msg.ProposalId)
+	}
+	if proposal.Status != govtypesv1.StatusVotingPeriod {
+		logger.Error(
+			"proposal is not in voting period",
+			"proposal", proposal.Id,
+			"status", proposal.Status,
+			"authority", msg.Vetoer,
+		)
+
+		return nil, errors.Wrapf(govtypes.ErrInactiveProposal, "proposal with ID %d is not in voting period", msg.ProposalId)
+	}
+
+	// follows the same logic as in x/gov/abci.go for rejected proposals
+	if msg.BurnDeposit {
+		ms.k.govKeeper.DeleteAndBurnDeposits(ctx, proposal.Id)
+	} else {
+		ms.k.govKeeper.RefundAndDeleteDeposits(ctx, proposal.Id)
+	}
+	proposal.Status = govtypesv1.StatusVetoed
+
+	// Since the proposal is veoted, we set the final tally result to an empty tally.
+	emptyTally := govtypesv1.EmptyTallyResult()
+	proposal.FinalTallyResult = &emptyTally
+
+	ms.k.govKeeper.SetProposal(ctx, proposal)
+	ms.k.govKeeper.RemoveFromActiveProposalQueue(ctx, proposal.Id, *proposal.VotingEndTime)
+	ms.k.govKeeper.DecrementActiveProposalsNumber(ctx)
```

Comment:
What about proposal votes ? It seems like they are going to stay in the store forever right ? I think we should remove them as well.